psi-sample/enclave/src/main/java/com/r3/conclave/sample/enclave/PSIEnclave.java

```diff
@@ -0,0 +1,117 @@
+package com.r3.conclave.sample.enclave;
+
+import com.esotericsoftware.kryo.Kryo;
+import com.esotericsoftware.kryo.io.Input;
+import com.r3.conclave.enclave.Enclave;
+import com.r3.conclave.mail.EnclaveMail;
+import com.r3.conclave.sample.common.AdDetails;
+import com.r3.conclave.sample.common.InputData;
+import com.r3.conclave.sample.common.InputDataSerializer;
+import com.r3.conclave.sample.common.UserDetails;
+
+import java.io.ByteArrayInputStream;
+import java.security.PublicKey;
+import java.util.*;
+import java.util.stream.Collectors;
+
+/**
+ * This class is the Enclave class. To create an enclave in Conclave extend your class by Enclave interface.
+ * This class will contain the our business logic to calculate the ad conversion rate by using the list of credit card numbers given by the merchant
+ * and the service provider. Since the lists are shared with the enclave and not with anyone else, complete privacy of data is guaranteed.
+ */
+public class PSIEnclave extends Enclave {
+
+    private static final String MERCHANT = "MERCHANT";
+    private static final String SERVICE_PROVIDER = "SERVICE-PROVIDER";
+    private List<UserDetails> userDetailsList;
+    private List<AdDetails> adDetailsList;
+
+    private Map<String , String> routingHintMap = new HashMap();
+    private Map<String, PublicKey> clientToPublicKey = new HashMap();
+    private Kryo kryo = new Kryo();
+
+
+    /**
+     * This method gets called when client wants to communicate to enclave, and sends a message wrapped in a mail to host. Host in turn calls deliverMail method which in turn
+     * calls this method. In this method, we will deserialize the mail message, perform the computation and send the result back to the clients.
+     * @param id
+     * @param mail
+     * @param routingHint
+     */
+    @Override
+    protected void receiveMail(long id, EnclaveMail mail, String routingHint) {
+
+        //deserialize the mail object using custom deserializers
+        InputData inputData = deserialize(mail);
+
+        //retrieve the envelope from mail. envelope is not encrypted
+        String envelope = new String(mail.getEnvelope());
```

Comment:
I think we need to indicate that the contents of the envelope tell us what type of client it is

Maybe we could do this by adding a comment and renaming `envelope` to `clientType`

Reply:
moved clientType to InputData.